File: AWS/iac-microbenchmark/ec2/python/rsa4096_encrypt.py
import base64
import json
import os
import sys
import secrets
import boto3
from botocore.exceptions import ClientError
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

# Initialize AWS KMS client
kms_client = boto3.client('kms')

def main():
    # Get the name argument from sys.argv
    request_json_raw = sys.argv[1]

    request_json = json.loads(request_json_raw)

    # Extract the message from the input payload
    message = request_json.get('message')

    # Get the KMS key ID from environment variables
    rsa_kms_key_id = os.environ['RSA4096_KMS_KEY_ARN']

    # Generate a random AES key and IV
    aes_key = secrets.token_bytes(32)
    iv = secrets.token_bytes(16)

    # Encrypt the data using AES
    cipher = Cipher(algorithms.AES(aes_key), modes.CTR(iv), backend=default_backend())
    encryptor = cipher.encryptor()
    plaintext = message.encode('utf-8')
    ciphertext = encryptor.update(plaintext) + encryptor.finalize()

    # Encrypt the AES key using KMS
    response = kms_client.encrypt(
        KeyId=rsa_kms_key_id,
        Plaintext=aes_key,
        EncryptionAlgorithm="RSAES_OAEP_SHA_256"
    )
    encrypted_aes_key = response['CiphertextBlob']

    # Return the results
    result = {
        'iv': base64.b64encode(iv).decode('utf-8'),
        'ciphertext': base64.b64encode(ciphertext).decode('utf-8'),
        'encrypted_aes_key': base64.b64encode(encrypted_aes_key).decode('utf-8')
    }

    print(json.dumps(result))

if __name__ == "__main__":
    main()

File: AWS/iac-microbenchmark/lambdas/typescript/rsa4096_encrypt.ts
import { KMSClient, EncryptCommand } from "@aws-sdk/client-kms";
import * as crypto from "crypto";
import { Context, APIGatewayProxyResult, APIGatewayEvent } from 'aws-lambda';

// Initialize KMS Client
const kmsClient = new KMSClient({});

export const handler = async (event: APIGatewayEvent, context: Context): Promise<APIGatewayProxyResult> => {

  // Get the KMS key ID from environment variables 
  const rsaKmsKeyId = process.env.RSA4096_KMS_KEY_ARN;

  if (!rsaKmsKeyId) {
    return {
      statusCode: 400,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "content-type": "application/json"
      },
      body: JSON.stringify({ message: "KMS Key ID not provided" }),
    };
  }

  try {
    // Generate a random AES key (256 bits)
    const aesKey = crypto.randomBytes(32);

    // Generate a random IV (16 bytes)
    const iv = crypto.randomBytes(16);

    // Create AES-CTR cipher
    const cipher = crypto.createCipheriv('aes-256-ctr', aesKey, iv);

    // Encrypt the message from the event
    const plaintext = Buffer.from(event.body ? JSON.parse(event.body).message : '', 'utf-8');
    const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);

    // Encrypt the AES key using KMS
    const encryptCommand = new EncryptCommand({
      KeyId: rsaKmsKeyId,
      Plaintext: aesKey,
      EncryptionAlgorithm: "RSAES_OAEP_SHA_256",
    });

    const response = await kmsClient.send(encryptCommand);
    const encryptedAesKey = response.CiphertextBlob;

    // Return encrypted data, IV, and AES key
    return {
      statusCode: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "content-type": "application/json"
      },
      body: JSON.stringify({
        iv: iv.toString('base64'),
        ciphertext: ciphertext.toString('base64'),
        encryptedAesKey: Buffer.from(encryptedAesKey!).toString('base64'),
      }),
    };
  } catch (error) {
    console.error("Error encrypting data: ", error);
    return {
      statusCode: 500,
      headers: {
        "Access-Control-Allow-Origin": "*",
        "content-type": "application/json"
      },
      body: JSON.stringify({ message: "Encryption failed" }),
    };
  }
};
